tweet_recommendations/other_methods/dbscan_based_method.py

```diff
@@ -0,0 +1,143 @@
+        # as many clusters as there are samples in dataset
+        neighbours_count = fit_params.get("neighbours_count", 256)
+
+        if self.verbose:
+            print(f"Data input shape: {x.shape}")
+
+        valid_hashtags = self.drop_tweets_with_hashtags_that_occurred_less_than(x,
+                                                                                minimal_hashtag_occurence)
+        x = self.drop_tweets_which_not_contain_given_hashtags(x, valid_hashtags)
+
+        if self.verbose:
+            print("Setup tweet embedding method")
+
+        self.embedding_method.fit(x)
+
+        if self.verbose:
+            print("Tweet content lemmas embedding started")
+
+        embeddings = self.embedding_method.transform(x["lemmas"])
+
+        if self.verbose:
+            print(f"Data embedding shape, after "
+                  f"droping data by given criteria: {embeddings.shape}")
+
+        self.neighbours = KDTree(embeddings)
+
+        epsilon = np.mean(self.neighbours.query(embeddings, k=neighbours_count, n_jobs=mp.cpu_count())[0])
+
+        if self.verbose:
+            print(f"Found epsilon: {epsilon}")
+            print("Fitting DBSCAN ... ")
+
+        self.clusterizer = DBSCAN(metric='manhattan', eps=epsilon, min_samples=MIN_SAMPLES, n_jobs=mp.cpu_count())
+        x["cluster_label"] = self.clusterizer.fit_predict(embeddings)
+
+        if self.verbose:
+            print("Clustering finished.")
+
+        x["hashtags"] = x["hashtags"].apply(lambda r: [elem["text"] for elem in r])
+
+        grouped_clusters = x.groupby(["cluster_label"])
+
+        if self.verbose:
+            print(f"Samples grouped into {len(np.unique(self.clusterizer.labels_))} clusters")
+            print("Building clusters' data")
+
+        self._centroids_data = []
+        self._corresponding_to_centroids_data_hashtags = []
+
+        for cluster_number, cluster in tqdm.tqdm(grouped_clusters, total=len(np.unique(x["cluster_label"])),
+                                                 disable=not self.verbose):
+            cluster_embeddings = np.vstack(cluster["embedding"].to_numpy())
+            cluster_center = np.mean(cluster_embeddings, axis=0).reshape(1, -1)
+            distances = distance.cdist(cluster_center, cluster_embeddings)[0]  # matrix of shape (1 x N) is returned
+
+            cluster_medoid_index = np.argmin(
+                distances)  # we are sure that it will not point to itself because mean center should be not existent
+
+            cluster_medoid = cluster_embeddings[cluster_medoid_index]
+            aggregated_hashtags = list(
+                chain.from_iterable(cluster["hashtags"].to_numpy()))
+            sorted_most_popular_hashtags = [hashtag for hashtag, count in
+                                            Counter(aggregated_hashtags).most_common()]
+
+            self._centroids_data.append(cluster_medoid)
```

Comment:
Jeśli dodajemy medoidę to chyba `_medoids_data` będzie trafniejsze

Reply:
No właśnie to jest moje rozjechanie zrozumienia, czym to jest. W publikacji odnoszą się do tego jako centroidy

Reply:
._.
Ale OK, to niechaj tak zostanie